### DISCUSSION TITLE
os.run 为什么只能在 target 域中使用

### DISCUSSION THREAD

**Discussion by osenberg-x** [xmake-io/xmake#4106] [Q&A]
### 你在什么场景下需要该功能？

想要在构建之前使用 os.run 执行一些命令，比如在构建之前 git clone 一个源码子项目
os.run("git clone -b url ${projectdir}/src/name")

### 描述可能的解决方案

可以在全局使用 os.run

### 其他信息

![image](https://github.com/xmake-io/xmake/assets/42266214/55227602-c48f-4602-a2eb-0eee9c32f41e)

[a comment by an automated account is not shown]

**Comment by waruqi** [comment]
先看下 描述域和脚本域的区别。。https://xmake.io/#/zh-cn/guide/syntax_description

os.run 只能在脚本域调用，全局定义 before_run 里面执行就行了。或者定义个 phony target 。其他 targets 都依赖它

**Comment by osenberg-x** [comment]
> 先看下 描述域和脚本域的区别。。https://xmake.io/#/zh-cn/guide/syntax_description
> 
> os.run 只能在脚本域调用，全局定义 before_run 里面执行就行了。或者定义个 phony target 。其他 targets 都依赖它

@waruqi 好的，明白了，但是还有一个疑问，文档里说在描述域里面一般不建议执行耗时高的操作，但是 os.run 应该很有可能做出耗时的操作啊?

[a comment by an automated account is not shown]

**Comment by waruqi** [comment]
> > 先看下 描述域和脚本域的区别。。https://xmake.io/#/zh-cn/guide/syntax_description
> > os.run 只能在脚本域调用，全局定义 before_run 里面执行就行了。或者定义个 phony target 。其他 targets 都依赖它
> 
> @waruqi 好的，明白了，但是还有一个疑问，文档里说在描述域里面一般不建议执行耗时高的操作，但是 os.run 应该很有可能做出耗时的操作啊?

所以 os.run 不是在描述域禁用的么，原本就不能用。感觉你还是没看明白什么是描述域。

**Reply by osenberg-x** [reply]
嗷，我描述的不够清楚，比如说在这个情况下
target("test")
  on_build(function (target)
    os.run("git clone -b master url $(projectdir)/src/name")
  end)

[a comment by an automated account is not shown]
